AP_NavEKF3: document provenance of Python-generated equations

tpwrules opened the pull request:
Over time the current iteration of the generator has drifted from the actual code, both due to changes within itself and in the used libraries. This PR tracks down how that's happened and makes everything consistent again so that informed improvements can be made.

* Code generator is modified to automatically use `sq` and friends.
* Old generator versions that generate the current code have been dug up.
* Generators ensure the correct version of SymPy is used (1.9)
* [Nix](https://nixos.org/) shell script added to conveniently use the correct versions of the generator libraries.
* Outdated previously generated code is removed from the repo.
* Unused generated code is disabled.

No compiler output change (on CubeOrange at least).

tridge commented:
I think this looks great, needs approval by @priseborough

rmackay9 commented:
LGTM

tridge reviewed (changes requested):
i think we should try to use one version of sympy for all generation, and check it generates good code, then assert sympy version instead of using nix-shell

Georacer reviewed (commented):
I'm interested to learn more about this part of our workflow.

I understand that these autogenerated files are meant to be copied manually into the `AP_NavEKF3` .cpp files and slightly modified? I'm saying this, as I don't see an exact 1:1 match of code-blocks between the files in `generated` and the library.

Do we consider this a drawback, as it means we don't fully automate the update procedure? Or is it desirable that it is that way?

tpwrules commented on libraries/AP_NavEKF3/derivation/main.py:
```diff
@@ -1,4 +1,10 @@
-#!/usr/bin/env python3
+#!/usr/bin/env nix-shell
+#! nix-shell --pure -i python3 -p "python3.withPackages (p: [ p.numpy p.sympy ])"
+#! nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/70bdadeb94ffc8806c0570eb5c2695ad29f0e421.tar.gz
+
+# above pins Python 3.10.13, Numpy 1.24.2, and Sympy 1.11.1 (and deps)
+# using the last nixos-23.05 branch commit
+
```
This means you can just run the `generate.sh` script and the environment will be automatically set up before the script starts and torn down once it ends. Now that I think about it I should be able to move the nix-shell to the shell script to avoid disrupting non-Nix users.

tpwrules commented on libraries/AP_NavEKF3/AP_NavEKF3_core.cpp:
```diff
@@ -2114,7 +2114,7 @@ void NavEKF3_core::calcTiltErrorVariance()
     const ftype &q3 = stateStruct.quat[3];
 
-    // equations generated by quaternion_error_propagation(): in AP_NavEKF3/derivation/main.py
+    // equations generated by quaternion_error_propagation(): in derivation/generate_2.py
```
It's in the implemented equations. Notice that some of the `PSn` variables are missing. Longer term we will just generate without having to modify. The difference is pretty obvious if you have the generated code.

tpwrules commented:
My goal with this PR was to minimize changes to the code and just document things as they are. Currently the results are indeed manually copied, and hopefully the steps should be relatively obvious now that the generated code matches what's actually being used.

We discussed in the dev call and indeed I think everyone would like to just run the generator and have the generated code committed (as the generator is pretty slow) and just include it. The current way is definitely not desirable, but improving it is down the road.

Georacer commented:
> My goal with this PR was to minimize changes to the code and just document things as they are. Currently the results are indeed manually copied, and hopefully the steps should be relatively obvious now that the generated code matches what's actually being used.
> 
> We discussed in the dev call and indeed I think everyone would like to just run the generator and have the generated code committed (as the generator is pretty slow) and just include it. The current way is definitely not desirable, but improving it is down the road.

Understood. Thanks for taking the time to answer!

tpwrules commented:
> i think we should try to use one version of sympy for all generation, and check it generates good code, then assert sympy version instead of using nix-shell

Moved the nix stuff to its own file and added the version asserts. Two different versions are still required. Would prefer to change the code in a separate PR, though I still need to examine how big the change will be.

tpwrules commented:
Found a version of Sympy that works for both scripts, so only one version needs to be installed. Both scripts check for that version.
